### PR TITLE
fix(training): prevent multi-GPU validation deadlock in COCO mAP sync

### LINKED ISSUE #931: Multi-GPU Validation fails
### Search before asking

- [x] I have searched the RF-DETR issues and found no similar bug report.

### Bug

NCCL timeout in multi-GPU training 

I trained for 2 epochs on single-GPU machine and then wanted to run the training on multi-GPU machine for faster training but I am seeing NCCL timeout after being stuck for a long time at the end of eval
```
Epoch 2: 100%|███████████████████████████████████████████████████████| 2289/2289 [19:11<00:00,  1.99it/s, train/lr=0.0001, train/lr_min=4.04e-6, train/lr_max=0.0001^
[  idation DataLoader 0: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 303/303 [02:01<00:00,  2.49it/s]
                                                                                                                                                                     

^[[A^[[A^[[A^[[A^[[A^[[A^[[A^[  [rank1]:[E407 10:02:26.551756518 ProcessGroupNCCL.cpp:683] [Rank 1] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=492
09, OpType=ALLGATHER, NumelIn=43, NumelOut=172, Timeout(ms)=1800000) ran for 1800056 milliseconds before timing out.                                                 
[rank1]:[E407 10:02:26.551997332 ProcessGroupNCCL.cpp:2241] [PG ID 0 PG GUID 0(default_pg) Rank 1]  failure detected by watchdog at work sequence id: 49209 PG status
: last enqueued work: 49209, last completed work: 49208                                                                                                              
[rank1]:[E407 10:02:26.552039573 ProcessGroupNCCL.cpp:730] Stack trace of the failed collective not found, potentially because FlightRecorder is disabled. You can en
able it by setting TORCH_NCCL_TRACE_BUFFER_SIZE to a non-zero value.                                                                                                 
[rank1]:[E407 10:02:26.552111664 ProcessGroupNCCL.cpp:2573] [PG ID 0 PG GUID 0(default_pg) Rank 1] First PG on this rank to signal dumping.                          
[rank3]:[E407 10:02:26.564157461 ProcessGroupNCCL.cpp:683] [Rank 3] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=49210, OpType=ALLREDUCE, NumelIn=1,
 NumelOut=1, Timeout(ms)=1800000) ran for 1800069 milliseconds before timing out.                                                                                    
[rank3]:[E407 10:02:26.564359555 ProcessGroupNCCL.cpp:2241] [PG ID 0 PG GUID 0(default_pg) Rank 3]  failure detected by watchdog at work sequence id: 49210 PG status
: last enqueued work: 49210, last completed work: 49209                                                                                                              
[rank3]:[E407 10:02:26.564374605 ProcessGroupNCCL.cpp:730] Stack trace of the failed collective not found, potentially because FlightRecorder is disabled. You can en
able it by setting TORCH_NCCL_TRACE_BUFFER_SIZE to a non-zero value
```

### Environment

- rfdetr=latest 
- num-gpus = 4
- Driver Version: 580.95.05
- CUDA Version: 13.0
- 2.9.1+cu128

### Minimal Reproducible Example

```py
from rfdetr import RFDETRBase

model = RFDETRBase()

model.train(
    dataset_dir="IE_Z22_043/images",
    epochs=200,
    batch_size=8,
    grad_accum_steps=4,
    task="segment",
    resolution=624,
    output_dir="rfdetr_ie_z22_043_finetuning",
    progress_bar=True,
    # fp16_eval=True,
    early_stopping=True,
    devices="auto",  # required — see note below
    strategy='ddp_find_unused_parameters_true',
    persistent_workers=True,
    device='cuda',
    checkpoint_interval=1,
    num_workers=4,
    warmup_epochs=5,
    pin_memory = True,
    resume="rfdetr_ie_z22_043_finetuning/checkpoint_1.ckpt"
)
```

I ran with `torchrun --nproc_per_node=4 train_rfdetr.py `

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes, I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Aaryan562**
## What does this PR do?

Fixes a **deadlock** in multi-GPU (DDP) **segmentation** validation. After an epoch, validation
hangs at the metric-sync step and is eventually killed by the NCCL watchdog:

[rankN] Watchdog caught collective operation timeout:
WorkNCCL(SeqNum=..., OpType=ALLGATHER, ...) ran for 1800000 milliseconds before timing out

It only reproduces **at scale** — a small/smoke validation set passes, the full set deadlocks —
which is what made it hard to catch.

### Background: how validation metrics are combined across GPUs

Under DDP, each GPU runs validation on its **own shard** of images. To produce one global mAP, the
per-GPU metric states must be **combined**, which `torchmetrics` does with a series of NCCL
collectives at `compute()` time. The hard rule of NCCL is that **every rank must issue the exact
same sequence of collectives, in the same order** — if one rank issues even one fewer/extra
collective, the ranks fall out of step and block on each other forever.

### Root cause: a divergent collective *count*, not a slow collective

This is a **circular-wait deadlock**, not a "slow" collective (so a longer NCCL timeout does not
help — it just delays the failure). The divergence is inside torchmetrics' own sync. Its
`gather_all_tensors` helper decides **how many collectives to run for each state tensor based on
that tensor's *local* `ndim`**:

- **scalar** (`ndim == 0`) → **1** collective (`all_gather` of the value)
- **vector** (`ndim ≥ 1`) → **2** collectives (`all_gather` of the *size*, then of the *data*)

`torchmetrics` only supports gathering tensors that have the **same number of dimensions on every
rank** — but DDP seg validation violates that, because different shards produce
differently-shaped state.

#### Worked example (4 GPUs)

Take one of MAP's per-class state tensors (e.g. scores for a given class). Suppose rank 3's shard
happened to contain no instances of that class this epoch, so its state collapses to a **0-dim
scalar**, while ranks 0/1/2 hold a normal **1-D vector**:

| | ranks 0, 1, 2 (vector) | rank 3 (scalar) |
|---|---|---|
| collectives issued for this state | `all_gather(size)` + `all_gather(data)` = **2** | `all_gather(value)` = **1** |

Rank 3 has now issued **one fewer** collective. NCCL matches collectives purely by issue order, so
from here on rank 3 is **off by one** — its next gather lines up with the others' current gather.
A few state tensors later, rank 3 is gathering its `labels` (int64) while ranks 0/1/2 are gathering
`scores` (float) **at the same sequence number** → the ops don't match → all ranks block → watchdog
fires after 30 min.

Captured live with `TORCH_DISTRIBUTED_DEBUG=DETAIL` (which turns the silent hang into an explicit
error):

Detected mismatch between collectives on ranks.
Rank 3:     CollectiveFingerPrint(SequenceNumber=40104, OpType=ALLGATHER, TensorShape=[1], TensorDtypes=Long)
Rank 0/1/2: CollectiveFingerPrint(SequenceNumber=40104, OpType=ALLGATHER, TensorShape=[1], TensorDtypes=Float)

This is a known, open `torchmetrics` limitation: Lightning-AI/torchmetrics#3199.

### Fix

Stop relying on torchmetrics' internal sync and **combine the per-rank state ourselves with a
single, fixed, shape-independent collective** — the exact pattern RF-DETR already uses for its F1
metric (`distributed_merge_matching_data`). torchmetrics still computes the mAP; we only replace
its one fragile merge step.

- Both mAP metrics are created with `sync_on_compute=False` (disables the internal sync).
- `_merge_metric_state_across_ranks()` gathers each metric's accumulated state with the repo's
  `all_gather` (a pickle-based object gather whose collective pattern is **identical on every rank
  regardless of tensor shape**), concatenates it, then `compute()` runs **locally** on the merged
  full-set state — producing the identical global mAP with no shape-dependent collective.
- The EMA metric is created **deterministically on every rank** (`on_validation_epoch_start`) so its
  gather is issued symmetrically too (a second, independent source of the same divergence).

**Why not the usual workaround?** The common suggestion (`sync_on_compute=False` +
`self.log(..., sync_dist=True)`) is mathematically wrong for mAP: `sync_dist` *averages* per-rank
values, but the mAP of the union of predictions ≠ the mean of per-shard mAPs.

**Related Issue(s):** Closes #931

## Type of Change

- Bug fix (non-breaking change that fixes an issue)

## Testing

- [x] I have tested this change locally
- [x] I have added/updated tests for this change

**Test details:**
- New CPU unit tests in `tests/training/test_coco_eval_callback.py`:
  - `_merge_metric_state_across_ranks` — no-op in single-process; concatenates each state list across ranks under a mocked 2-rank gather.
  - EMA-metric symmetry — created on `on_validation_epoch_start` for every rank regardless of batch count; the unanimous `all_reduce(MIN)` gate decides EMA compute identically on all ranks.
- Full CPU suite (`pytest -m "not gpu"`) and `pre-commit run --all-files` pass.
- **4× L4, full dataset, end-to-end:** DDP validation now completes with no collective
  mismatch/deadlock and produces correct merged mAP/segm/F1 (the previous code reproducibly crashed
  at this exact step, as captured above).

### CODE REVIEW

**Comment by Aaryan562** [comment]
@Borda pls review!

**Comment by Borda** [comment]
> [CLA assistant check](https://cla-assistant.io/roboflow/rf-detr?pullRequest=1085) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/roboflow/rf-detr?pullRequest=1085) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/roboflow/rf-detr?pullRequest=1085) it.

@Aaryan562 could you please sign or share printsceen that you already did? :raccoon:

**Comment by Aaryan562** [comment]
> > [CLA assistant check](https://cla-assistant.io/roboflow/rf-detr?pullRequest=1085) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/roboflow/rf-detr?pullRequest=1085) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/roboflow/rf-detr?pullRequest=1085) it.
> 
> @Aaryan562 could you please sign or share printsceen that you already did? 🦝

image
I have already signed, but its not showing, so attached the pic @Borda

**Comment by Aaryan562** [comment]
@Borda pls let me know the further steps or clarification required. Happy to share, contribute and learn along the way.

**Comment by isaacrob** [comment]
cool find. why not try to fix this upstream in torchmetrics, as it seems like a bug in their code?

**Comment by Borda** [comment]
> > > [CLA assistant check](https://cla-assistant.io/roboflow/rf-detr?pullRequest=1085) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/roboflow/rf-detr?pullRequest=1085) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/roboflow/rf-detr?pullRequest=1085) it.
> > 
> > @Aaryan562 could you please sign or share printsceen that you already did? 🦝
> 
> image
> I have already signed, but its not showing, so attached the pic @Borda 

Perfect, thank you!

**Comment by Aaryan562** [comment]
> cool find. why not try to fix this upstream in torchmetrics, as it seems like a bug in their code?

Thanks, @isaacrob ! The issue has been open for over a year now, and there hasn't been much activity on it from the maintainers' side.

For now, I'd like to address it within RF-DETR, especially since multi-GPU training issues for segmentation models exists from v1.6.2. We've been using RF-DETR for a few of our use cases, and training on a single GPU with a dataset of around 60k samples takes roughly 4 hours per epoch. That was the primary motivation behind contributing this fix.

**Comment by isaacrob** [comment]
Got it, makes sense!

When we did the original pretraining for this model on many GPUs we used a different codebase that did not use lightening, this version isn't as battle tested at scale I guess. Nice find

**Comment by Borda** [comment]
> cool find. why not try to fix this upstream in torchmetrics, as it seems like a bug in their code?

Yeah, would be great, just not sure if it would land and being released any time soon...

**Review by Borda — DISMISSED** [review]

**Review by Borda — APPROVED** [review]

**Comment by Aaryan562** [comment]
@Borda Thanks for merging the PR! I'm curious which release/version this fix is expected to be included in.
